### PR TITLE
dt: child binding compatibles should match their parents

### LINKED ISSUE #29758: edtlib not reporting proper matching_compat for led nodes (and other children nodes)
For a DTS like (from frdm_k64f.dts):

```
        leds {
                compatible = "gpio-leds";
                red_led: led_0 {
                        gpios = < &gpiob 0x16 0x1 >;
                        label = "User LD1";
                };
                green_led: led_1 {
                        gpios = < &gpioe 0x1a 0x1 >;
                        label = "User LD2";
                };
                blue_led: led_2 {
                        gpios = < &gpiob 0x15 0x1 >;
                        label = "User LD3";
                };
        };
```

the `matching_compat` for `/leds/led_0` should be `gpio-leds` but its being reported as `None`

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Fixes: #29758

See commit log in the second patch for details.

### CODE REVIEW

File: scripts/dts/edtlib.py [hunk 1]
```diff
@@ -1508,6 +1508,14 @@ def __init__(self, path, fname2path, raw=None,
             if key.endswith("-cells"):
                 self.specifier2cells[key[:-len("-cells")]] = val
 
+        # Make child binding compatibles match ours if they are missing.
```
Comment (galak) [inline scripts/dts/edtlib.py:1511]: would it make sense to do this after line(s) 1492? where we initially set self.child_binding?

File: scripts/dts/edtlib.py [hunk 2]
```diff
@@ -1508,6 +1508,14 @@ def __init__(self, path, fname2path, raw=None,
...
+        if self.compatible is not None:
+            child = self.child_binding
+            while child is not None:
```
Comment (galak) [inline scripts/dts/edtlib.py:1514]: curious about the use of `while` here?

File: scripts/dts/edtlib.py (hunk 1 above)
Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:1511]: > would it make sense to do this after line(s) 1492? where we initially set self.child_binding?

I think we want to do this after 1500 at least (where we call `self._check`) to make sure we're doing it on a well-formed binding. Doing it after initializing the look up tables means the parent i fully initialized before we go and fix up the child, which seems like a nice thing to have done in case we need to fix up the child look up tables based on the parent eventually, etc.

Happy to move it after self._check though if you want.

**Review by galak — COMMENTED** [review]
LGTM, just a few style questions on the code.

File: scripts/dts/edtlib.py (hunk 2 above)
Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:1514]: > curious about the use of `while` here?

`child-binding:` can recurse

**Review by galak — APPROVED** [review]

File: scripts/dts/edtlib.py (hunk 1 above)
Comment (galak) [reply scripts/dts/edtlib.py:1511]: Fine as, is just curious.

**Comment by mbolivar-nordic** [comment]
Weird macos error:

```
ERROR: Could not find a version that satisfies the requirement wheel (from versions: none)
ERROR: No matching distribution found for wheel
```

File: scripts/dts/edtlib.py [hunk 3]
```diff
@@ -1390,9 +1390,9 @@ class Binding:
       The free-form description of the binding.
 
     compatible:
-      The compatible string the binding matches. This may be None if the
-      Binding object is a child binding or if it's inferred from node
-      properties.
+      The compatible string the binding matches. If the Binding object
+      is parsed from a child binding, this may be the 'compatible' value of the
```
Comment (pabigot) [inline scripts/dts/edtlib.py:1394]: "...from a child binding *that does not have a `compatible` property*, this may be..."

Or something that makes it clear that if the child has a compatible it does not *have* to be equal to the parent.

**Comment by galak** [comment]
> Weird macos error:
> 
> ```
> ERROR: Could not find a version that satisfies the requirement wheel (from versions: none)
> ERROR: No matching distribution found for wheel
> ```

weid, going to try re-running the check and see what happens.

**Review by pabigot — CHANGES_REQUESTED** [review]
I'd like clarification of the behavior when the child binding already has a compatible that is not the same as its parent.

On the overall concept of inferring a compatible for the child please see https://github.com/zephyrproject-rtos/zephyr/issues/29758#issuecomment-721253772.  This does not make sense to me.

**Comment by galak** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/29763 Fixes the build issue that this uncovers.

File: scripts/dts/edtlib.py (hunk 3 above)
Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:1394]: > Or something that makes it clear that if the child has a compatible it does not _have_ to be equal to the parent.

Sure, I've tried to reword this to make that more clear. I added a test as well.

**Comment by mbolivar-nordic** [comment]
> On the overall concept of inferring a compatible for the child please see [#29758 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/29758#issuecomment-721253772). This does not make sense to me.

I'm a bit confused why that is a blocker for the PR, which is restoring old behavior, not introducing new behavior?

**Comment by pabigot** [comment]
> I'm a bit confused why that is a blocker for the PR, which is restoring old behavior, not introducing new behavior?

It's not a blocker.  You replied to my request for documentation clarification, but I don't see a change...?

File: scripts/dts/edtlib.py (hunk 3 above)
Comment (pabigot) [reply scripts/dts/edtlib.py:1394]: I think my problem is with "if condition, then this may happen", when I'd like to see more clearly what *will* happen in different conditions.

**Comment by mbolivar-nordic** [comment]
>  You replied to my request for documentation clarification, but I don't see a change...?

D'oh. I copy/pasted advice from git that recommended `git push --set-upstream origin dt-child-binding-compatible` and accidentally created a branch in zephyrproject-rtos. I really need to stop doing that. Sorry.

**Review by pabigot — APPROVED** [review]
